Add wxConvAuto::GetEncoding(), other improvements and fixes

pavel-t opened the pull request:
Add functions to check the encoding that wxConvAuto is using. Also fix a few bugs and add tests.

pavel-t committed:
Add wxConvAuto::IsFallbackEncoding()

pavel-t committed:
Add wxConvAuto::IsUTF8()

pavel-t committed:
Add wxConvAuto::GetEncoding()

pavel-t committed:
Check wxConvAuto state in tests

pavel-t committed:
Add more wxConvAuto test cases

pavel-t committed:
Fix wxTextInputStream incorrectly decoding multibyte fallback encodings

pavel-t committed:
Do not delete and create the fallback conversion again when it fails

pavel-t committed:
Switch to fallback earlier if the input is not valid UTF-8 prefix

pavel-t committed:
Fix wxTextInputStream for some inputs starting with nulls

vadz commented on include/wx/convauto.h:
```diff
@@ -91,6 +93,14 @@ class WXDLLIMPEXP_BASE wxConvAuto : public wxMBConv
+    wxFontEncoding GetEncoding() const;
+
+    // Return true if the fall-back encoding is used
+    bool IsFallbackEncoding() const
```
This name seems a bit confusing to me, I'd rather call it `IsUsingFallbackEncoding()` or `UsesFallbackEncoding()`.

I also wonder if it's really that useful -- couldn't the application code do `GetEncoding() == GetFallbackEncoding()` if it really needs to know?

pavel-t replied:
If's useful for objects constructed with custom fallback encoding -- otherwise it would need something like GetBOM() == wxBOM_None && !IsUTF8(). I'll rename it.

vadz commented on include/wx/strconv.h:
```diff
@@ -387,6 +387,8 @@ class WXDLLIMPEXP_BASE wxMBConvUTF8 : public wxMBConvStrictUTF8
     int m_options;
 };
 
+bool wxIsUTF8Prefix(const char *src, size_t len);
```
Ideally we shouldn't put this private function in a public header. The trouble is that we don't seem to have any appropriate file in `include/wx/private` where we could put it and creating a new header just for a single function doesn't seem appealing... But I guess we still should do it, `include/wx/private/unicode.h` maybe?

If not, we should at the very least put it into `wxPrivate` namespace but, again, ideally we wouldn't have it in the public header at all.

Oh, and I'd also rename this one to `wxCanBeUTF8Prefix()`, but this is much less important for a private function.

pavel-t replied:
I only put it in `strconv.cpp` to use `tableUtf8Lengths[]`, so maybe we should move this table to private header? This would also allow to remove its copy from `ustring.cpp`.

vadz replied:
I had no idea we had a copy of it there to be honest... We definitely shouldn't have it there, but I don't think we should define this array (`tableUtf8Lengths`) in the header, we'd have to make it static and it still would be duplicated. We could declare it there, of course, and still keep it defined in `strconv.cpp` and I'd be glad if you could modify it like this, of course. TIA!

vadz commented on src/common/convauto.cpp:
```diff
@@ -307,22 +307,19 @@ wxConvAuto::ToWChar(wchar_t *dst, size_t dstLen,
         // decoding UTF-8 input when fed it byte by byte, as done by
         // wxTextInputStream, for example
-        if ( srcLen < m_conv->GetMaxCharLen() )
+        if ( srcLen < 2 + m_conv->GetMaxCharLen() && wxIsUTF8Prefix(src, srcLen) )
```
Could you please add a comment here explaining this `+ 2`? I don't really understand where does it come from and what would break if we didn't have it.

vadz reviewed (commented):
Thanks a lot for the updates and the tests! There were clearly quite a few bugs here, sorry about that and thanks for fixing them.

We can merge this almost as it is, but I'd like to either rename or (even simpler) get rid of `IsFallbackEncoding()` (but perhaps I'm missing why it's more useful than I think). What do you think?

pavel-t committed:
Rename IsFallbackEncoding() to IsUsingFallbackEncoding()

pavel-t committed:
Add comment about 2 extra bytes in wxConvAuto::ToWChar

pavel-t committed:
Fix GetEncoding() after global fallback encoding is changed

Save global fallback encoding when switching to it, so GetEncoding() can still return the correct value if global fallback encoding is changed later. Also do not switch to wxFONTENCODING_MAX when it is set as global fallback encoding.

pavel-t committed:
Add wx/private/unicode.h

pavel-t committed:
Move wxIsUTF8Prefix() to convauto.cpp

pavel-t committed:
Use tableUtf8Lengths[] in sringops.cpp too

pavel-t committed:
Make leading nulls a special case to avoid breaking decoding some short strings in fallback encoding

pavel-t committed:
Allow decoding even shorter strings in fallback encoding

Complete UTF-8 characters (except leading nulls) never appear in failed decoding attempts when the input is fed byte by byte.

pavel-t committed:
Fix UTF-8 build

vadz commented:
Thanks again! Will merge soon.
